### DISCUSSION TITLE
It takes ~2 minutes to open a folder with ~3000 files, is it normal or something not correct?

### DISCUSSION THREAD

**Discussion by happyringtail** [Studio-42/elFinder#3326] [Q&A]
Hi guys,

I use elfinder with laravel and aws S3, it takes  ~2 minutes to open a folder within 3000 files. Is it normal speed or something goes wrong? I tried 'checkSubfolders' => -1 in config, but it doesn't help.

It seems the bottleneck is studio-42/elfinder/php/elFinderVolumeDriver.class.php, it takes ~0.8s to load file list from s3, but ~120s to finish this loop.

```
foreach ($this->scandirCE($path) as $p) {
            if (($stat = $this->stat($p)) && empty($stat['hidden'])) {
                if (!$hasDir && $stat['mime'] === 'directory') {
                    $hasDir = true;
                }
                $this->dirsCache[$path][] = $p;
            }
        }
```

Is there anything I can do to optimize this function or some special configs I can us in config file?

Thank you so much!

**Comment by nao-pon** [comment]
Since elFinder needs to check all attributes such as read/write files, it makes an inquiry for each file. The following code from barryvdh/elfinder-flysystem-driver will be called, and I think you'll get better performance if you can optimize this part.

https://github.com/barryvdh/elfinder-flysystem-driver/blob/5a6c893dfb97e9848d7b1e5e990e943af7bc3550/src/Driver.php#L232-L333

**Comment by liquidaty** [comment]
This issue is not purely back-end; at the least, in 'list' view, when scrolling through a folder with more than a few hundred files, the front-end will also significantly lag. As far as I can tell, it is rendering each entry as its own TR (vs paginating across a fixes set of TR elements), which is a fairly well-known problem for browser performance. I can appreciate that changing the table rendering to use an alternative higher-performing approach is not a trivial task, but it would be a *huge* improvement to the front-end performance even with folders with as few as one or two hundred rows, and we would be willing to contribute in the form of a bounty if that can help to improve front-end performance.
